### DISCUSSION TITLE
Compiling Gleam to serializable js

### DISCUSSION THREAD

**Discussion by nadavhames** [gleam-lang/gleam#2875] [Ideas & suggestions]
Lets say I have a record in Gleam like this:
```gleam
pub type Person {
	Teacher(name: String, age: String)
	Human(name: String)
	Me(last_name: String)
}
```
...and I compile it to js. The output I get uses classes:
```ts
export class Teacher extends _.CustomType {
  constructor(name: string, age: string);
  
  name: string;
  age: string;
}

export class Human extends _.CustomType {
  constructor(name: string);
  
  name: string;
}

export class Me extends _.CustomType {
  constructor(last_name: string);
  
  last_name: string;
}

export type Person$ = Teacher | Human | Me;
```
This is fine, but what if my js codebase requires serializability of data/data structures? Classes in js afaik are not serializable. One idea to make a serializable structure would be to have a serializable flag in the js target config which tells the compiler to emit pure js objects instead of classes. It could look something roughly like this:
```ts
export interface Teacher {
 typeTeacher: true;
 name: string;
 age: string;
}
export interface Human {
 typeHuman: true;
 name: string;
}
export interface Me {
 typeMe: true;
 last_name: string;
}
export type Person = Teacher | Human | Me;
```
For each of the Teacher, Human, and Me types, helper functions could also be emitted for easy object creation and also for identity checking. 
Alternatively, it might be nice to have a helper that comes with the emitted js using the class based setup that can serialize and deserialize the class structure into objects.
Would be interested to hear what insights people have on this topic!

**Comment by coder0107git** [comment]
`interface` only exists in Typescript, not vanilla js. Also, `interface`s don't get emmitted during transpilation to js. One option would be to add a `toJSON` or `toString` method to the compiled classes or to `_.CustomType`.

**Comment by samifouad** [comment]
in javascript classes are a thin abstraction on the underlying object prototype system... you should be able to just use built-in features when trying to bridge your existing JS project with your gleam compiled output, unless i'm misunderstanding what you're trying to accomplish

[link](https://jsfiddle.net/s6wdm72j/):
![Screenshot from 2024-03-30 23-00-52](https://github.com/gleam-lang/gleam/assets/6378290/83bd6f09-0449-4d73-90dd-5e898cb6e17c)

**Reply by coder0107git** [reply]
What's the `Object.assign({}, source)` for? It works fine without that.

**Reply by samifouad** [reply]
yeah, was just showing off how it's all interchangable with a language built-in
